### ISSUE TITLE
TEdit got NIL charlooks while editing message in Lafite send window

### ISSUE THREAD

**Issue by nbriggs** [Interlisp/medley#1018]
I was deleting text to get a message ready to send, and TEdit broke with "ARG NOT CHARLOOKS" (NIL) attempting to \TEDIT.APPLY.STYLES under \TEDIT.CHARDELETE

I don't know if I can reproduce this at will, it was fiddly deleting characters to get the message to look right.  Initially the message was plain text, but then I shift-selected a piece of formatted text into the message, and started to delete characters ahead of that.

![Screen Shot 2022-11-06 at 1 04 25 PM](https://user-images.githubusercontent.com/8138841/200196068-d5410090-c641-4466-82a1-0edb655ffca0.png)

**Comment by rmkaplan** [comment]
I assume this is in my new TEDIT branch.

I don’t think I am yet supporting Lafite, it has some low-level dependencies that I don’t totally understand.  And styles are separate issue.  As far as I can tell, they are not documented anywhere, not clear where they come from and where they go.  Lafite may provide some insight into that.

But the call from \TEDIT.CHARDELETE should not be passing a NIL piece, maybe that’s the source, no matter what styles are doing. If you have a real PC, then it should have real looks.

**Comment by nbriggs** [comment]
this was with TEdit dated 26-Oct-2022, TEDIT.;105

**Comment by nbriggs** [comment]
I'm trying to duplicate the problem, and I've got it in a very confused state.  As I delete characters in the middle of the message it changes a character displayed at the end of the message.  I've attached a tedit-bug-lisp.virtualmem.gz.   You should be able to bring the Lafite sender to the top, and if you type a character where the cursor is (in front of "guest") you should see that character appear there and at the end of the message.
[tedit-bug-lisp.virtualmem.gz](https://github.com/Interlisp/medley/files/9946909/tedit-bug-lisp.virtualmem.gz)

**Comment by rmkaplan** [comment]
I’ll look.  But, have you ever been able to create/send a Lafite message with the new code?

**Comment by nbriggs** [comment]
Yes, plain text only, but I presume it’s still using tedit to do it

**Comment by nbriggs** [comment]
Hmmm... I was running with Nethub networking -- I just checked and it will hang if you don't have a Dodo network running for it to attach to (-nh-host localhost).   If you do that, and it asks for a login, give it guest/guest as name/password.

Yes, I was using the delete key.
I was trying to delete a character to the left of the caret, I had probably moved there from somewhere else, it was probably a left button selection.
The character I was trying to delete was probably at the beginning of a line, adjacent to the headers, not at the beginning or end of a paragraph or the whole text.

I can restart my saved version ONLY if

**Comment by rmkaplan** [comment]
How did you start up the message?  Maybe I can replicate in my environment.  I assume I load Lafite the old fashioned way, then what?

What happens if you click on a character in a header?  Does it let you edit there or slide over to the right?

**Comment by rmkaplan** [comment]
On loading LAFITE:  I have an old init file that loads it and sets some variables.  It calls (LAFITEMODE ‘UNIX), but that gets an invalid arg error.  Maybe because LAFITEMODELST is NIL.  (Also wants to know my NSNAME, I assume that doesn’t matter for this kind of test).

**Comment by nbriggs** [comment]
I built my maiko with  `-DMAIKO_ENABLE_NETHUB` added to the OPTFLAGS in `maiko/bin/makefile-darwin.x86_64-x`.  (from maiko/bin:  `./makeright x cleanup` and `./makeright init` and `./makeright x`)

I did a `git clone https://github.com/devhawala/dodo.git` and then in the `dodo` directory, `unzip dist.zip`, which will give you a `dist` directory... `cd dist` in two different terminal windows: `./run-nethub.sh` in the first, and `./run-dodoserver.sh` in the second.

I started a full sysout, `./run-medley -nh-host localhost -full` and then in an Interlisp exec `(\NSINIT)` -- it takes about 3-4 seconds to start up before you can type that at the exec.  If `(CH.FINDSERVER)` returns an NS address it's working properly.

I connect to `{...}...>library>lafite>`, then `(FILESLOAD LAFITE NEWNSMAIL)`, then when it's all loaded `(LAFITEMODE 'NS)` and `(LAFITE :ON)`

From the Lafite control window, click on "Send Mail"...

**Comment by nbriggs** [comment]
I tried without the Dodo networking, and as long as you ignore the break windows for the NS login, you can get to the point where you "Send Message" and start messing about with it.  This time I got an error extending a selection.

**Comment by rmkaplan** [comment]
I found that if I load UNIXMAIL, then it finds the right mode, brings up the lafite window, and I can bring up a standard form form by clicking “Send Mail”.  Since I don’t really want to send mail, just see what Tedit does, I think that’s enough for now.

What I see so far in the is that the header fields are not protected (you can edit “Subject:” to anything you like).  There are EOL’s but no paragraph breaks, just one long piece with EOL’s spread around.  

The only odd thing is that it seems to end with a zero-length empty piece, after the >>Message<<[EOL].  Should be OK (The old and new logic is supposed to skip empty pieces and non-empty pieces that are marked as “invisible"), but maybe not.

Otherwise, it looks like a simple plain-text stream, nothing fancy.  Fiddling around with random inserts and deletes, I haven’t gotten it to break.

It may have one glitch, in that it doesn’t seem to include (i.e. underline) the EOL’s in the selection when you click beyond the end of the line.  So to delete it, you have to click at the beginning of the next line, and backspace.  I’ll have to check whether that’s a change in behavior from the old scanning code. I had a lot of trouble understanding how it was accounting for EOL’s.
